### PR TITLE
[git-webkit] Handle force pushed branch in checkout

### PR DESCRIPTION
**Pull request by briannafan**
#### 62e8794cc8f57c914afa1146ab1e3520915a7572
<pre>
[git-webkit] Handle force pushed branch in checkout
<a href="https://bugs.webkit.org/show_bug.cgi?id=260874">https://bugs.webkit.org/show_bug.cgi?id=260874</a>
rdar://114656165

Reviewed by Jonathan Bedard.

Resets the head of tracking branch to match remote.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git.checkout):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/checkout.py:
(Checkout.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLand.test_default):
(TestLand.test_canonicalize):
(TestLand.test_default_with_radar):
(TestLand.test_canonicalize_with_bugzilla):
(TestLandGitHub):

Canonical link: <a href="https://commits.webkit.org/268933@main">https://commits.webkit.org/268933@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c8e350da4787bb9b41cab6f9d8ba228c46706234

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21110 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22988 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21350 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/24740 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21334 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/24740 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/18305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/23842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/21298 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/24740 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/19135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/23842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/24740 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/19336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/23842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/19880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/21672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/20931 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2608 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19734 "Built successfully") | | |

### CODE REVIEW

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 1]
```diff
@@ -1032,8 +1032,27 @@ def checkout(self, argument, prune=None):
...
             ).returncode else self.commit()
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
-        if branch_remote:
+
+        # if branch has a remote and is not a dev branch
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            sys.stderr.write("Resetting branch {} to remote {}.\n Checkout will erase all un-pushed changes on {}.\n".format(argument, path, argument))
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: We generally reserve `stderr` for something going wrong, I don't think that's the case here. This seems like `log.info`, or maybe even `log.debug`.

As a general statement, `print` is reserved for required output of scripts (like `git-webkit pr` outputting the URL of the posted PR). `sys.stderr` is reserved for errors that we expect users to react to. `log.*` is for everything else (and notably, controlled by `-q` and `-v`, which isn't the case for `print` and `sys.stderr`)

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 2]
```diff
@@ -1032,8 +1032,27 @@ def checkout(self, argument, prune=None):
...
+            if Terminal.choose(
+                'Ensure all desired changes have been saved. Continue with checkout?\n',
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I would have this say:
`"'{}' looks like a PR branch. Would you like to override the local version of this branch with the version from '{}'".format(argument, branch_remote)`

Comment (briannafan) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: As discussed, not relevant as PR branches are filtered out. (Also, significant changes in amended commit incoming.)

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 3]
```diff
@@ -1032,7 +1032,47 @@ def checkout(self, argument, prune=None):
                 cwd=self.root_path,
             ).returncode else self.commit()
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
+
+        # if branch is not a dev branch and has a remote
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: We should start a comment like this with 'if', instead, should be something like:
`Branch is not a dev branch and exists on a remote`

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 4]
```diff
@@ -1032,7 +1032,47 @@ def checkout(self, argument, prune=None):
...
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            local_head = self.commit(branch=argument)
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: We can be a bit more efficient here because we only really need the `hash`, we don't care about the identifier or commit log (which are both a bit more computationally expensive to fetch):
`self.commit(branch=argument, include_log=False, include_identifier=False)`

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 5]
```diff
@@ -1032,7 +1032,47 @@ def checkout(self, argument, prune=None):
...
+                    capture_output=True,
+                    encoding='utf-8',
+                ).stdout.strip()
+                branches_for_local_1 = self.branches_for(local_head_1_hash)
+
+                # check if head~1 has remote branch, in case of local commit
+                if set(branches_for_local_1) != set(branches_for_remote):
+                    log.info("Local branch is not tracking the remote branch.")
+                    if Terminal.choose(
+                        "Local changes on {} will not be saved. Would you like to override the local version of this branch with the version from '{}'?".format(argument, path),
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:1119]: We should guard this prompt with a flag that is `false` by default, and have `git-webkit checkout` pass the `true`. This function is potentially used in circumstances where the user wouldn't expect to be prompted, perhaps because the "user" is actually a machine.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 6]
```diff
@@ -1032,7 +1032,47 @@ def checkout(self, argument, prune=None):
...
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
+
+        # if branch is not a dev branch and has a remote
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            local_head = self.commit(branch=argument)
+            remote_head = self.commit(branch=path)
+            branches_for_local = self.branches_for(local_head.hash)
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I don't think this quite gives us the answer we need. We only want to prompt the user if the user is at risk of losing data. The only way the user is at risk of losing data is if the ref they're targeting includes a change which doesn't exist on any remote. At first glance, this seems like a challenging problem to efficiently solve, I though it was `O(n)` where `n` is the number of remote branches we track (which in WebKit, may be in the low thousands).

However, @gsnedders showed me `git merge-base` a few months back, and that turns out to basically magically answer our question here.

If you invoke a command like this:
```
git merge-base <hash> remotes/<remote>/<branch-a> remotes/<remote>/<branch-b>...
```
you will end up with a hash. If that hash is equal to the `<hash>` you passed in as the first argument, then the provided hash is a member of at least one of the provided remote branches. Otherwise, you know that <hash> represents a commit which only exists locally.

I think this is the approach you want to take here. `git merge-base` runs extremely quickly , your only catch is that Windows has a limit of 32K characters in a single command, so you want to limit the number of arguments to your shell invocation. `git-webkit branch` (which uses `git merge-base` to achieve a similar goal) just caps the number of arguments at 512, which seems like a prudent approach here as well.

Comment (gsnedders) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I lack some of the context here, given I don't know when we're force pushing to the tracked remote.

Do we expect the former head of the tracked remote to continue to be on some other branch, because if not we could potentially end up with a very large number of orphaned commits which probably are of no interest to the author?

When it comes to `git merge-branch`, we should probably refactor the code that's currently in `program.branch.Branch.branch_point` to be somewhere else (in https://github.com/WebKit/WebKit/pull/13934#discussion_r1203069779 I seemingly suggested moving it to `Repository`, so maybe let's do that?).

That said, I don't know how common this workflow is, nor how much we should care about this.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py (hunk 3 above)
Comment (gsnedders) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: Also WebKit code style is full sentences (thus ending in a ~~full stop~~ period).

**Review by gsnedders — COMMENTED** [review]
Not really a. comment on this PR specifically, but I can't help but feel like we're just ended up with way, way, way too much in the `checkout` function.

It feels kinda hard to follow what each of the different cases actually is, and reason about why we have different behaviour for each case.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py (hunk 6 above)
Comment (briannafan) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I think my previous implementation did give the right behaviour but your approach is more efficient, so I've used `branch_point`. I decided to keep the checks for what branches are associated with the commit, as I don't want to reset the branch if it isn't necessary. So, if there's a local change but the local branch is tracking the correct remote branch, a normal checkout is performed.

Comment (briannafan) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: #18445

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 7]
```diff
@@ -1094,7 +1094,45 @@ def checkout(self, argument, prune=None):
...
+
+        # Branch is not dev and exists on a remote.
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
+            local_head = self.commit(branch=argument, include_log=False, include_identifier=False)
+            remote_head = self.commit(branch=path, include_log=False, include_identifier=False)
+            branches_for_remote = self.branches_for(remote_head.hash)
+            local_head_1_ref = '{}~1'.format(argument)
+            local_head_1_hash = run(
+                [self.executable(), 'rev-parse', local_head_1_ref],
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I don't understand why you are using `{argument}~1` here instead of just `{argument}`?

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 8]
```diff
@@ -1094,7 +1094,45 @@ def checkout(self, argument, prune=None):
...
+                cwd=self.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            ).stdout.strip()
+
+            # Resets branch if local is not tracking force pushed remote.
+            branches_for_local_1 = self.branches_for(local_head_1_hash)
+            if set(branches_for_local_1) != set(branches_for_remote):
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I think this logic is more complicated than it needs to be. I believe this is as simple as:
```
hsh = run(
    [self.executable(), 'rev-parse', local_head_1_ref],
    cwd=self.root_path,
    capture_output=True,
    encoding='utf-8',
).stdout.strip()[:Commit.HASH_LABEL_SIZE]
bp = self.branch_point(ref= hsh)
if bp.hash != hsh:
    log.info(" You have unsaved changes on the local branch.")
    ...
```

Put another way "Could I make a pull-request from the ref you provided? No? Ok, then you can't possibly have changes you care about associated with it"

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py (hunk 7 above)
Comment (briannafan) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: This is a good question. I think I confused myself by thinking about it too hard.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py (hunk 8 above)
Comment (briannafan) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: I was mainly thinking that we don't want to reset the branch unnecessarily if it's already tracking the remote. So by checking the branches associated with the commit, we can see if the remote is tracked. Only if it isn't do we need to check for local changes and reset the branch. Agree with the `branch_point` logic for prompting though. 
```
branches_for_local_1 = self.branches_for(local_head_1_hash)
if set(branches_for_local_1) != set(branches_for_remote):
     branch_point = self.branch_point(ref=local_head.hash)
     if branch_point != local_head.hash:
          [prompt in terminal]
[perform normal checkout without reset]
```

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 9]
```diff
@@ -1094,7 +1094,42 @@ def checkout(self, argument, prune=None):
                 cwd=self.root_path,
             ).returncode else self.commit()
 
+        match = self.dev_branches.match(argument)
         branch_remote = self.remote_for(argument)
+
+        # Branch is not dev and exists on a remote.
+        if not match and branch_remote:
+            path = '{}/{}'.format(branch_remote, argument)
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: Nit: I think we want a better variable name than "path" here

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 10]
```diff
@@ -1094,7 +1094,42 @@ def checkout(self, argument, prune=None):
...
+            local_head = self.commit(branch=argument, include_log=False, include_identifier=False)
+            remote_head = self.commit(branch=path, include_log=False, include_identifier=False)
+            local_bp = self.branch_point(ref=local_head.hash)
+            merge_base = run(
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: Can we call this "merge_base_with_target_remote" or something like that? This code has a lot of magic in it, so I think we need to rely heavily on verbose variable names to explain what's going on.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 11]
```diff
@@ -1094,7 +1094,42 @@ def checkout(self, argument, prune=None):
...
+                cwd=self.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            ).stdout.strip()
+
+            # Resets branch if local is not tracking force pushed remote.
+            if merge_base != local_bp.hash:
+                if local_bp.hash != local_head.hash:
+                    log.info(" You have unsaved changes on the local branch.")
+                    if prompt:
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: You can combine this pair into a single `if prompt and Terminal.choose(...` to be less verbose.

**Review by JonWBedard — APPROVED** [review]
